### PR TITLE
bridge, session: Drop changing umask

### PR DESCRIPTION
**Pull request by martinpitt**
cockpit-session set a tight 077 umask for extra paranoia. However, on    
OSes without pam_umask configured by default (like Debian), this causes    
all session processes to run with that umask as well. This then caused    
e.g. `cockpit.file('/etc/motd').replace()` to write motd with 0600    
permissions, making it inaccessible. (cf. issue #18033)    
    
The C bridge counteracted this by harcoding `umask(022)`, but that    
really does not belong there: The bridge also gets run through SSH or    
directly from ws. The Python bridge does no such thing, and we rather    
don't want to add that.    
    
There is no real benefit for cockpit-session to change the umask. It    
does not create any files by itself, and security related files created    
by GSSAPI or PAM modules should already do their own explicit permission    
setting instead of relying on a tight umask. For example, krb5's    
open_file() ensures 0600 permissions when writing ccache files. (The    
only case where a new file is plausibly written during login).    

----

This fixes [most of the mess here](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18757-20230508-070615-e88af50c-debian-testing-bots-4733/log.html#98-2), see https://github.com/cockpit-project/cockpit/pull/18757#issuecomment-1538015838 for the debugging details.

### CODE REVIEW

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
I think the "restore the umask" part, at least, is an outright bug, due to its apparent ability to override the user's chosen umask via `pam_umask`.  Why isn't this working normally?  Is there some PAM function we need to call to make sure the postlogin stuff gets executed?

I'd actually like to see if we could think about not mucking with umask at all from `cockpit-session`...

File: src/bridge/bridge.c [hunk 1]
```diff
@@ -368,9 +368,6 @@ run_bridge (const gchar *interactive,
         g_warning ("couldn't change to runtime dir: %s: %s", directory, g_strerror (errno));
     }
 
-  /* Reset the umask, typically this is done in .bashrc for a login shell */
-  umask (022);
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/bridge/bridge.c:372]: This is actually a good point. Given the test failures, it is actually *certain* that cockpit would break some things such as /etc/motd permissions if the OS uses a tight 077 umask by default. We are once again at the point where issue #18033 bites us.

So from a "clean room" perspective, I'd like to drop all this umask mucking around (in both the bridge and c-session), and make cockpit.file() respect existing file permissions, or go over all API users and explicitly specify owner and permissions (which is sometimes annoying and race-y to figure out from JS, though). But as we don't really know who else uses this API, it now seems to me that this is water under the bridge, and we just need to do that umask thing in the Py bridge as well. Would that be acceptable?

File: src/session/session-utils.c [hunk 2]
```diff
@@ -580,6 +580,9 @@ spawn_and_wait (const char **argv, const char **envp,
       if (n_remap_fds != -1)
         fd_remap (remap_fds, n_remap_fds);
 
+      /* restore original umask */
+      umask (session_umask);
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/session/session-utils.c:584]: Debian doesn't use `pam_umask` in its default PAM config. The only reference is in /etc/login.defs:
```
# UMASK is the default umask value for pam_umask and is used by
# useradd and newusers to set the mode of the new home directories.
# 022 is the "historical" value in Debian for UMASK
# 027, or even 077, could be considered better for privacy
# There is no One True Answer here : each sysadmin must make up his/her
# mind.
UMASK           022
```

But without pam_umask this feels like a no-op. It seems the expected place to set this is ~/.profile, see /etc/skel/.profile:
```
# the default umask is set in /etc/profile; for setting the umask
# for ssh logins, install and configure the libpam-umask package.
#umask 022
```

So I don't really see anything which would affect the umask for a session, that's why we see the 077 survive. This feels unsatisfactory, but that's something which we need to live with somehow -- AFAICS, either we drop umask() from cockpit-session, or we add it to the Python bridge.

File: src/session/session.c [hunk 3]
```diff
@@ -914,8 +915,8 @@ main (int argc,
 
   program_name = basename (argv[0]);
 
-  /* Cleanup the umask */
-  umask (077);
+  /* Tight umask during authentication */
+  mode_t orig_umask = umask (077);
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/session/session.c:919]: I checked: The only direct writing open() call that cockpit-session does is this:

    open (_PATH_BTMP, O_WRONLY | O_APPEND);

However:

 - it usually exists already, hence the umask/mode shouldn't matter; when it's deleted, /usr/lib/tmpfiles.d/var.conf restores the permissions on the next boot.
 - the call does *not* specify `O_CREAT`, so if it does not exist, it would just get an ENOENT
 - the file has root:utmp 0660 permissions, so creating it with either 077 or 022 umasks would both be wrong. 

So I think this call is not relevant, after all.

The only other place that could write files that I can think of is if GSSAPI writes a kerberos ticket cache file (yes, it normally goes to the kernel keyring or the daemon, but `$KRB5CCNAME` is still a thing). World-readable files would indeed be disastrous there. But [krb5](https://github.com/krb5/krb5/blob/master/src/lib/krb5/rcache/rc_dfl.c#L103) does the right thing already and open these with a `0600` mask (it has to, otherwise every `kinit` invocation would be vulnerable).

So in summary, I don't see anything which would go wrong with 022 permissions.

**Comment by martinpitt** [comment]
@allisonkarlitskaya : I replied to the threads above. I didn't push any changed code, let's first find an agreement. Thanks!

File: src/session/session-utils.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/session/session-utils.c:584]: I wonder if we ought to be invoking `cockpit-bridge` via the shell... we've gone back and forth on this a number of times... this feels like an argument in favour of doing it again.

File: src/session/session.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/session/session.c:919]: I think we should just leave the umask alone, then.

**Comment by martinpitt** [comment]
Summary from IRC discussion: Let's avoid hacks to counteract other hacks. Drop the umask mangling from both c-session and c-bridge. Do *not* add libpam-umask to cockpit.pam in Debian, as (1) that's not how Debian is meant to configure umasks, and (2) it would actually be detrimental: As our code relies on 022 for historical reasons, we should not actually configure a possibly changed umask from /etc/login.defs, but just keep the 022 default.

**Review by allisonkarlitskaya — APPROVED** [review]
Looks great!

Ship it!  :rocket:
